text/0000-canonical-metadata.md

```diff
@@ -0,0 +1,1198 @@
+          // refId to check the name of the binding in the source module
+          name: string;
+          moduleSpecifier: string;
+          refId: ID; // Back pointer to the metadata for external module reference
+          location: SourceLocation;
+      }
+    | 'unresolved'; // When the parent class is not statically analyzable, for example mixins
+
+// Information about a module reference
+interface ModuleReference {
+    // Full canonical name, eg. lightning/input, lightning/recordForm
+    // name will be in camel case
+    moduleSpecifier: string;
+    // name of the bundle or module, will retain camel case
+    name?: string;
+    // namespace if it exists
+    namespace?: string;
+    // Only present for salesforce scoped modules, else property is omitted
+    sfdcResource?: {
+        scoped:
+            | 'accessCheck'
+            | 'apex'
+            | 'apexContinuation'
+            | 'client'
+            | 'community'
+            | 'component'
+            | 'contentAssetUrl'
+            | 'customPermission'
+            | 'dynamicComponent'
+            | 'messageChannel'
+            | 'i18n'
+            | 'gate'
+            | 'label'
+            | 'metric'
+            | 'internal'
+            | 'resourceUrl'
+            | 'schema'
+            | 'slds'
+            | 'user'
+            | 'userPermission';
+        namespaceId?: string; // scoped module namespace normalization
+    };
+    type:
+        | 'lwc' // reference is the built-in lwc module
+        | '@salesforce' // reference is a @salesforce scoped module
+        | 'internal' // reference is relative import, "internal" to the current bundle
+        | 'external'; // reference is an external module like another component or a library
+    location: SourceLocation;
+    id: ID;
+}
+
+// Information about default function parameters.
+// Also look at parameters represented using rest syntax.
+interface DefaultParameters {
+    type: 'DefaultParameters';
+    name: string;
+    defaultValue?: number | string | boolean | null | object;
+}
+
+// Information about function parameters received using the rest(...) syntax.
+interface RestParameters {
+    type: 'RestParameters';
+    name: string;
+    startIndex: number;
+    endIndex?: number;
+}
+
+// Unique identifier per object, other metadata objects of the same module can have a back reference
+// to this id. The uniqueness of ids is only guaranteed within the module. It cannot span multiple modules
+type ID = string;
+```
+Examples:
+Samples of the metadata for script files can be viewed [here](https://github.com/salesforce/lwc-metadata/pull/2)
+
+#### CSS file metadata shape
+```js
+// Root metadata object for a css file
+interface CSSFile {
+    fileType: 'css';
+    fileName: string;
+    imports?: ModuleReference[];
+    customProperties?: CSSCustomProperty[];
+    staticResources?: StaticResourceReference[];
+}
+
+// Information about css custom property.
+interface CSSCustomProperty {
```

Comment:
Do we need a property to distinguish between declaration and usage?

Reply:
The shape has been updated. Declarations and references are called out separately.

Reply:
@ravijayaramappa Any reason this distinction between reference and declaration was removed?